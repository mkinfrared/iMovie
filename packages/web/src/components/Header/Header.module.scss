@import "src/styles/colors";
@import "src/styles/animations";

.Header {
  color: #000000;

  .menu {
    margin-right: 12px;
    padding: 12px;
    @extend .colorTransition;
  }

  .toolbar {
    align-items: center;
    display: flex;

    > a {
      align-items: center;
      color: inherit;
      display: flex;
      height: 1.6em;
      text-decoration: none;
      @extend .colorTransition;

      > svg {
        height: 100%;
      }

      > h6 {
        margin-left: 10px;
      }
    }

    .loginButton {
      margin-left: auto;
      @extend .colorTransition;
    }
  }
}
