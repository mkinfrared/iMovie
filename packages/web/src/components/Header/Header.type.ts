import { ReactNode } from "react";

export interface HeaderProps {
  children?: ReactNode;
}

export interface Navigation {
  text: string;
  icon: JSX.Element;
}
