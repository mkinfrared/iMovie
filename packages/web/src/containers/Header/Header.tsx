import React from "react";

import HeaderFC from "components/Header";
import useAuthenticationCheck from "hooks/useAuthenticationCheck";
import useIsAdmin from "hooks/useIsAdmin";

const Header = () => {
  const isAuth = useAuthenticationCheck();
  const isAdmin = useIsAdmin();

  return <HeaderFC isAuth={isAuth} isAdmin={isAdmin} />;
};

export default React.memo(Header);
