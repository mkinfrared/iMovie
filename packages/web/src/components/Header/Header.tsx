import AppBar from "@material-ui/core/AppBar";
import Button from "@material-ui/core/Button";
import IconButton from "@material-ui/core/IconButton";
import Toolbar from "@material-ui/core/Toolbar";
import Typography from "@material-ui/core/Typography";
import MenuIcon from "@material-ui/icons/Menu";
import React, { useCallback, useState } from "react";
import { Link } from "react-router-dom";

import AdminMenu from "components/AdminMenu";
import Menu from "containers/Menu";
import Imovie from "icons/imovie";
import loadableModal from "utils/loadable";

import css from "./Header.module.scss";
import { HeaderProps } from "./Header.type";

const Login = loadableModal(() => import("containers/Login"));
const SignUp = loadableModal(() => import("components/SignUp"));

type ActiveForm = "signin" | "signup" | null;

const Header = ({ isAuth }: HeaderProps) => {
  const [isAdminOpen, setIsAdminOpen] = useState(false);
  const [activeForm, setActiveForm] = useState<ActiveForm>(null);

  const openDrawer = useCallback(() => {
    setIsAdminOpen(true);
  }, []);

  const closeDrawer = useCallback(() => {
    setIsAdminOpen(false);
  }, []);

  const openLogin = useCallback(() => {
    setActiveForm("signin");
  }, []);

  const openSignUp = useCallback(() => {
    setActiveForm("signup");
  }, []);

  const closeForms = useCallback(() => {
    setActiveForm(null);
  }, []);

  const renderButton = () => {
    if (isAuth) {
      return <Menu className={css.loginButton} />;
    }

    return (
      <Button
        color="inherit"
        className={css.loginButton}
        onClick={openLogin}
        aria-label="login"
      >
        Login
      </Button>
    );
  };

  const renderForm = () => {
    switch (activeForm) {
      case "signin":
        return (
          <Login
            open={activeForm === "signin"}
            onClose={closeForms}
            openSingUp={openSignUp}
          />
        );
      case "signup":
        return (
          <SignUp
            open={activeForm === "signup"}
            onClose={closeForms}
            openLogin={openLogin}
          />
        );
      default:
        return null;
    }
  };

  return (
    <>
      <AppBar position="static" className={css.Header}>
        <Toolbar className={css.toolbar}>
          <IconButton
            edge="start"
            className={css.menu}
            color="inherit"
            aria-label="menu"
            onClick={openDrawer}
          >
            <MenuIcon />
          </IconButton>
          <Link to="/">
            <Imovie />
            <Typography variant="h6">iMovie</Typography>
          </Link>
          {renderButton()}
        </Toolbar>
      </AppBar>
      <AdminMenu isOpen={isAdminOpen} onClose={closeDrawer} />
      {renderForm()}
    </>
  );
};

export default React.memo(Header);
