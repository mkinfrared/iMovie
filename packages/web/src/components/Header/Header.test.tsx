import { fireEvent, render } from "@testing-library/react";
import React, { ReactElement } from "react";
import { MemoryRouter } from "react-router-dom";

import Header from "./Header";

jest.mock("containers/Login", () => (props: any) => (
  <div data-testid="containers/Login">
    <code>{JSON.stringify(props)}</code>
  </div>
));

jest.mock("components/SignUp", () => (props: any) => (
  <div data-testid="containers/SignUp">
    <code>{JSON.stringify(props)}</code>
  </div>
));

jest.mock("components/AdminMenu", () => (props: any) => (
  <div data-testid="components/AdminMenu">
    <code>{JSON.stringify(props)}</code>
  </div>
));

jest.mock("components/Menu", () => (props: any) => (
  <div data-testid="components/Menu">
    <code>{JSON.stringify(props)}</code>
  </div>
));

describe("<Header />", () => {
  let Component: ReactElement;

  const isAuth = false;

  beforeEach(() => {
    Component = (
      <MemoryRouter>
        <Header isAuth={isAuth} />
      </MemoryRouter>
    );
  });

  it("should be defined", () => {
    const { container } = render(Component);

    expect(container).toBeDefined();
  });

  it("should match a snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should render Menu when user is authenticated", () => {
    const { getByTestId } = render(
      <MemoryRouter>
        <Header isAuth />
      </MemoryRouter>
    );

    const menu = getByTestId("components/Menu");

    expect(menu).toBeInTheDocument();
  });

  it("should open a sidebar", () => {
    const { getByLabelText, getByTestId } = render(Component);
    const menuButton = getByLabelText("menu");

    fireEvent.click(menuButton);

    const admin = getByTestId("components/AdminMenu");
    const result = JSON.parse(admin.textContent as string);

    expect(result.isOpen).toBe(true);
  });

  it("should open a Login component", async () => {
    const { getByLabelText, findByTestId } = render(Component);
    const loginButton = getByLabelText("login");

    fireEvent.click(loginButton);

    const login = await findByTestId("containers/Login");

    expect(login).toBeInTheDocument();
  });
});
