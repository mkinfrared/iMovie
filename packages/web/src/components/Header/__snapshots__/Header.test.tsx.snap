// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should match a snapshot 1`] = `
<div>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorSecondary Header MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular toolbar MuiToolbar-gutters"
    >
      <button
        aria-label="menu"
        class="MuiButtonBase-root MuiIconButton-root menu MuiIconButton-colorInherit MuiIconButton-edgeStart"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <a
        href="/"
      >
        <svg
          viewBox="0 0 512.01 512.01"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            iMovie
          </title>
          <circle
            cx="256"
            cy="256"
            fill="#f57c00"
            r="256"
          />
          <path
            d="M256,79.78h0L215.65,208.17l-134.57-1.3,101.7,101.7L147.89,412.5h0l99.34,99.33c2.91.1,5.83.17,8.77.17,131.17,0,239.29-98.67,254.22-225.83l-79.3-79.3-47.37.46Z"
            fill="#e65100"
          />
          <polygon
            fill="#fff"
            points="256 79.78 296.35 208.17 430.92 206.87 321.28 284.92 364.11 412.5 256 332.35 147.89 412.5 190.72 284.92 81.08 206.87 215.65 208.17 256 79.78"
          />
          <rect
            fill="#f57c00"
            height="53.04"
            width="59.38"
            x="219.87"
            y="229.48"
          />
          <rect
            fill="#f57c00"
            height="12.54"
            width="6.7"
            x="210.14"
            y="238.56"
          />
          <polygon
            fill="#f57c00"
            points="282.48 248.56 282.48 263.44 301.86 276.25 301.86 235.75 282.48 248.56"
          />
        </svg>
        <h6
          class="MuiTypography-root MuiTypography-h6"
        >
          iMovie
        </h6>
      </a>
      <div
        data-testid="containers/Menu"
      >
        <code>
          {"className":"loginButton"}
        </code>
      </div>
    </div>
  </header>
  <div
    data-testid="components/AdminMenu"
  >
    <code>
      {"isOpen":false}
    </code>
  </div>
</div>
`;
